pkg/libdevfile/libdevfile.go

```diff
@@ -69,6 +87,211 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+	if debugCmdErr == nil && !reflect.DeepEqual(emptyCommand, debugCommand) {
+		isDebugCommandValid = true
+		klog.V(2).Infof("Debug command: %v", debugCommand.Id)
+	}
+
+	if !isDebugCommandValid {
+		commandErrors := ""
+		if debugCmdErr != nil {
+			commandErrors += debugCmdErr.Error()
+		}
+		return v1alpha2.Command{}, fmt.Errorf(commandErrors)
+	}
+
+	return debugCommand, nil
+}
+
+// ValidateAndGetTestDevfileCommands validates the test command
+func ValidateAndGetTestDevfileCommands(data data.DevfileData, devfileTestCmd string) (testCommand v1alpha2.Command, err error) {
+	var emptyCommand v1alpha2.Command
+	isTestCommandValid := false
+	testCommand, testCmdErr := GetTestCommand(data, devfileTestCmd)
+	if testCmdErr == nil && !reflect.DeepEqual(emptyCommand, testCommand) {
+		isTestCommandValid = true
+		klog.V(2).Infof("Test command: %v", testCommand.Id)
+	}
+
+	if !isTestCommandValid && testCmdErr != nil {
+		return v1alpha2.Command{}, testCmdErr
+	}
+
+	return testCommand, nil
+}
+
+// GetBuildCommand iterates through the components in the devfile and returns the build command
+func GetBuildCommand(data data.DevfileData, devfileBuildCmd string) (buildCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileBuildCmd, v1alpha2.BuildCommandGroupKind)
+}
+
+// GetDebugCommand iterates through the components in the devfile and returns the debug command
+func GetDebugCommand(data data.DevfileData, devfileDebugCmd string) (debugCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileDebugCmd, v1alpha2.DebugCommandGroupKind)
+}
+
+// GetRunCommand iterates through the components in the devfile and returns the run command
+func GetRunCommand(data data.DevfileData, devfileRunCmd string) (runCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileRunCmd, v1alpha2.RunCommandGroupKind)
+}
+
+// GetTestCommand iterates through the components in the devfile and returns the test command
+func GetTestCommand(data data.DevfileData, devfileTestCmd string) (runCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileTestCmd, v1alpha2.TestCommandGroupKind)
+}
+
+// getCommand iterates through the devfile commands and returns the devfile command associated with the group
+// commands mentioned via the flags are passed via commandName, empty otherwise
+func getCommand(data data.DevfileData, commandName string, groupType v1alpha2.CommandGroupKind) (supportedCommand v1alpha2.Command, err error) {
+
+	var command v1alpha2.Command
+
+	if commandName == "" {
+		command, err = getCommandFromDevfile(data, groupType)
+	} else if commandName != "" {
+		command, err = getCommandFromFlag(data, groupType, commandName)
+	}
+
+	return command, err
+}
+
+// getCommandFromDevfile iterates through the devfile commands and returns the command associated with the group
+func getCommandFromDevfile(data data.DevfileData, groupType v1alpha2.CommandGroupKind) (supportedCommand v1alpha2.Command, err error) {
```

Comment:
I'm not sure about not having  named return values. 
I remember we discussed not to have implcit return statements (a return statement without indicating the values to return)

Reply:
> I remember we discussed not to have implcit return statements (a return statement without indicating the values to return)

Implicit return statement is a result of having a named return values.

Reply:
Having named return values is necessary to make implicit return, but you can have named return values and use explicit return.

IMO, named return values adds information to the signature and it would be nice to use it when having some bool or other non auto-documented value in the return signature

Reply:
> Having named return values is necessary to make implicit return, but you can have named return values and use explicit return.

Then this is what I am seeking.